docs/spec/consensus/checkvalidators.md

```diff
@@ -0,0 +1,224 @@
+
+  * *oh*: the old trusted header
+  * *nh*: the new header that has to be checked
+
+Let's assume $oh.bfttime + tp > nh.bfttime$ and $oh.bfttime + tp > now$.
+In the following examples, the pairs $(v,p)$ denote validators and their voting power.
+
+####Example: Identical VSets
+
+\[
+oh.V = \{(1,1), ... (4,1)\}\\
+nh.V = \{(1,1), ... (4,1)\}
+\]
+
+As we trust oh.V (at oh.bfttime) and the trusting period is not over yet, we
+trust nh.V.
+
+####Example: Changed Voting powers
+
+\[
+oh.V = \{(1,1), ... (4,1)\}\\
+nh.V = \{(1,1), ... (4,2)\}
+\]
+
+
+Validator 4 has more than a third voting power in nh.V. As trusting oh does not
+rule out that 4 is faulty, the fault assumption might be violated in $nh.V$. Thus, $nh.V$
+cannot be trusted.
+
+####Example: Lucky case with $n> 3t +1$
+
+\[
+oh.V = \{(1,1), ... (6,1)\}\\
+nh.V = \{(1,1), ... (7,1)\}
+\]
+
+  By the fault assumption ($n > 3t$), at most one validator in
+  $oh.V$ is faulty. In addition, validator 7 may be faulty. As a result
+  there are at most 2 faulty validators in $nh.V$. Because $7 > 3 \cdot 2$ we
+  say that oh.V provides sufficient trust in order to trust $nh.V$.
+
+####Example: Swapping validators
+
+\[
+oh.V = \{(1,1), ... (4,1)\}\\
+nh.V = \{(2,1), ... (5,1)\}
+\]
+
+
+
+
+Observe that validator 1 is not present in $nh.V$. Conservatively, we have to
+assume 1 is correct, and there may be a fault among 2,3,4. In addition, we don't
+know 5, so that conservatively, we have to assume 5 may be faulty. Thus among
+2,3,4,5, there may be two faults which violates the faults assumption. Thus $oh$
+does **not** provide sufficient trust in order to trust $nh$.
+
+
+## Basics for the "trusting period" method
+
+The function `CheckVS(oh, nh)` returns true, when *oh* provides sufficient
+trust to trust nh.
+
+### Assumptions
+
+1. $tp < unbonding period$.
+2. $nh.bfttime < now$
+3. $nh.bfttime < oh.bfttime+tp$
+4. $trust(oh)=true$
+
+### Some Maths
+
+**Observation 1.** If $oh.bfttime + tp > now$, we trust the old
+validator set $oh.V$.
+
+In the following let's assume oh is trusted and sufficiently new.
+
+**Definition 1.** Let $PA \subseteq oh.V$ be a *potential adversary* in $oh$, if
+the sum of the voting powers in PA is less than 1/3 of the voting
+powers in $oh.V$, that is,
+\[
+\sum_{(v,p) \in PA} p < 1/3 \sum_{(v,p) \in oh.V} p
+\]
+
+**Proposition 1.** The set of faulty processes
+\[oh.V \setminus \{(v,p): (v,p) \in oh.V \wedge correct(v,oh.bfttime + tp)\}\]
+is a potential adversary.
+
+*Proof.* By the LightClient invariant.
+
+**Definition 2.** Let the *unknown validators* UV be the validators that appear in $nh.V$ and not in
+$oh.V$, that is,
+\[
+UV = \{(v,p): (v,p) \in nh.V \wedge \nexists (v,x) \in oh.V \}.
+\]
+
+
+
+
+**Theorem 1.** If for all potential adversaries PA, in $nh$ the combined voting
+powers of PA and UV is less than a third of the total voting power, then in
+$nh$, more than 2/3 of the voting power is held by correct processes. Formally,
+if for all PA
+
```

Comment:
Note that in the expression below, we fix PA from old header, but we take into account only its voting power in the new header.

Reply:
But the calculation on the old header just gives us the maximum possible bad voting power. It doesn't tell us that in fact the validators in the PA that gives that maximum are faulty. It can happen that actual bad validators outside PA have increased in power and good validators have decreased in power. I will try to find an example for this.

Reply:
We don't do calculation on the old header. Note that the sum is only on p which is voting power of validators from PA in the new header. We only intersect with the old header.